### PR TITLE
External database access unification (jdbc/odbc bridges)

### PR DESCRIPTION
**Pull request by alex-krash**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

The purpose of this PR is to implement connect to external databases via odbc/jdbc bridges, spawned as separate daemons.
I am working in parallel on JDBC-bridge here - https://github.com/alex-krash/clickhouse-jdbc-bridge

The final result of this PR would be like:
1. Table functions "odbc" and "jdbc" exists
2. Engines "ODBC" and "JDBC" exists
3. (optional) support for JDBC dictionaries implemented

While working on this PR, I've figured out following problems, which I would like to solve as well:
1. Existing ODBC bridge forces using of double-quotes as escape sequence for identifiers. It breaks compatibility with MySQL, particularly. I propose to add API method to odbc/jdbc bridge, that will return value of escape character (or add it to existsing /column_info handler, and rename it)
2. Nullability support abscence. For JDBC bridge I've implemented it (see below), maybe we should implement it for ODBC as well?

Example of jdbc interaction with MySQL:
```
mysql> CREATE TABLE `test`.`test` (
    ->   `int_id` INT NOT NULL AUTO_INCREMENT,
    ->   `int_nullable` INT NULL DEFAULT NULL,
    ->   `float` FLOAT NOT NULL,
    ->   `float_nullable` FLOAT NULL DEFAULT NULL,
    ->   `double` DOUBLE NOT NULL,
    ->   `double_nullable` DOUBLE NULL DEFAULT NULL,
    ->   `timestamp` TIMESTAMP NOT NULL,
    ->   `timestamp_nullable` TIMESTAMP NULL DEFAULT NULL,
    ->   `date` DATE NOT NULL,
    ->   `date_nullable` DATE NULL DEFAULT NULL,
    ->   PRIMARY KEY (`int_id`));
Query OK, 0 rows affected (0,09 sec)

mysql> insert into test (`int_id`, `float`, `double`, `timestamp`, `date`) VALUES (1,2,3, '2018-01-02 03:04:05', '2019-05-06');Query OK, 1 row affected (0,00 sec)

mysql> select * from test;
+--------+--------------+-------+----------------+--------+-----------------+---------------------+--------------------+------------+---------------+
| int_id | int_nullable | float | float_nullable | double | double_nullable | timestamp           | timestamp_nullable | date       | date_nullable |
+--------+--------------+-------+----------------+--------+-----------------+---------------------+--------------------+------------+---------------+
|      1 |         NULL |     2 |           NULL |      3 |            NULL | 2018-01-02 03:04:05 | NULL               | 2019-05-06 | NULL          |
+--------+--------------+-------+----------------+--------+-----------------+---------------------+--------------------+------------+---------------+
1 row in set (0,00 sec)
```
From clickhouse-client:
```
DESCRIBE TABLE jdbc('jdbc:mysql://localhost:3306/?user=root&password=root', 'test', 'test')

┌─name───────────────┬─type───────────────┬─default_type─┬─default_expression─┐
│ int_id             │ Int32              │              │                    │
│ int_nullable       │ Nullable(Int32)    │              │                    │
│ float              │ Float32            │              │                    │
│ float_nullable     │ Nullable(Float32)  │              │                    │
│ double             │ Float64            │              │                    │
│ double_nullable    │ Nullable(Float64)  │              │                    │
│ timestamp          │ DateTime           │              │                    │
│ timestamp_nullable │ Nullable(DateTime) │              │                    │
│ date               │ Date               │              │                    │
│ date_nullable      │ Nullable(Date)     │              │                    │
└────────────────────┴────────────────────┴──────────────┴────────────────────┘

10 rows in set. Elapsed: 0.031 sec.

SELECT *
FROM jdbc('jdbc:mysql://localhost:3306/?user=root&password=root', 'test', 'test') 

┌─int_id─┬─int_nullable─┬─float─┬─float_nullable─┬─double─┬─double_nullable─┬───────────timestamp─┬─timestamp_nullable─┬───────date─┬─date_nullable─┐
│      1 │         ᴺᵁᴸᴸ │     2 │           ᴺᵁᴸᴸ │      3 │            ᴺᵁᴸᴸ │ 2018-01-02 03:04:05 │               ᴺᵁᴸᴸ │ 2019-05-06 │          ᴺᵁᴸᴸ │
└────────┴──────────────┴───────┴────────────────┴────────┴─────────────────┴─────────────────────┴────────────────────┴────────────┴───────────────┘

1 rows in set. Elapsed: 0.055 sec.
```

### CODE REVIEW

**Comment by alex-krash** [comment]
Currently, new implementation of ODBC, via unified bridge interface is registered via "idbc" name, in order to being able to perform comparison between them.

**Comment by alexey-milovidov** [comment]
> I propose to add API method to odbc/jdbc bridge, that will return value of escape character

Ok.

> maybe we should implement it for ODBC as well?

Yes. There is no reason, why it is not implemented.

> Currently, new implementation of ODBC, via unified bridge interface is registered via "idbc" name, in order to being able to perform comparison between them.

You can replace old ODBC to your implementation as soon as integration test passes.

File: dbms/programs/odbc-bridge/getIdentifierQuote.cpp [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+
+namespace DB
+{
+    std::string getIdentifierQuote(SQLHDBC hdbc)
+    {
+        std::string identifier = "";
+
+        SQLSMALLINT t;
+        SQLRETURN r = POCO_SQL_ODBC_CLASS::SQLGetInfo(hdbc, SQL_IDENTIFIER_QUOTE_CHAR, NULL, 0, &t);
```
Comment (alesapin) [inline dbms/programs/odbc-bridge/getIdentifierQuote.cpp]: Actually we know that the size of SQL_IDENTIFIER_QUOTE_CHAR is one or zero bytes. So maybe here we can allocate buffer (identifier) and request ODBC only once?

File: dbms/src/Common/XDBCBridgeHelper.h [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+struct JDBCBridgeMixin
+{
+    static constexpr inline auto DEFAULT_PORT = 9019;
+    static const String configPrefix() { return "jdbc_bridge"; }
+    static const String serviceAlias() { return "clickhouse-jdbc-bridge"; }
+    static const String getName() { return "JDBC"; }
+
+    static void startBridge(const Poco::Util::AbstractConfiguration & , const Poco::Logger * , const Poco::Timespan & )
+    {
+        throw Exception("jdbc-bridge is not running. Please, start it manually");
```
Comment (alesapin) [inline dbms/src/Common/XDBCBridgeHelper.h]: Error code

File: dbms/src/Dictionaries/XDBCDictionarySource.cpp [hunk 3]
```diff
@@ -57,54 +59,54 @@ namespace
 static const size_t max_block_size = 8192;
 
 
-ODBCDictionarySource::ODBCDictionarySource(const DictionaryStructure & dict_struct_,
+XDBCDictionarySource::XDBCDictionarySource(const DictionaryStructure & dict_struct_,
     const Poco::Util::AbstractConfiguration & config, const std::string & config_prefix,
-    const Block & sample_block, const Context & context)
-    : log(&Logger::get("ODBCDictionarySource")),
+    const Block & sample_block, const Context & context, const BridgeHelperPtr bridge)
```
Comment (alesapin) [inline dbms/src/Dictionaries/XDBCDictionarySource.cpp]: Constructor params shoud be named with underline `bridge_`

File: dbms/src/Dictionaries/XDBCDictionarySource.cpp [hunk 4]
```diff
@@ -57,54 +59,54 @@ namespace
...
+    : log(&Logger::get(bridge->getName() + "DictionarySource")),
     update_time{std::chrono::system_clock::from_time_t(0)},
     dict_struct{dict_struct_},
     db{config.getString(config_prefix + ".db", "")},
     table{config.getString(config_prefix + ".table")},
     where{config.getString(config_prefix + ".where", "")},
     update_field{config.getString(config_prefix + ".update_field", "")},
     sample_block{sample_block},
-    query_builder{dict_struct, db, table, where, IdentifierQuotingStyle::None},    /// NOTE Better to obtain quoting style via ODBC interface.
+    query_builder{dict_struct, db, table, where, bridge->getIdentifierQuotingStyle()},
```
Comment (alesapin) [inline dbms/src/Dictionaries/XDBCDictionarySource.cpp]: Seems like method `getIdentifierQuotingStyle()` called before `startBridge()`?

File: dbms/src/Common/XDBCBridgeHelper.h [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+    }
+
+    virtual ~XDBCBridgeHelper() {}
+
+    String getName() const override
+    {
+        return BridgeHelperMixin::getName();
+    }
+
+    IdentifierQuotingStyle getIdentifierQuotingStyle() override
```
Comment (alesapin) [inline dbms/src/Common/XDBCBridgeHelper.h]: Maybe startBridgeSync() should be called here?

File: dbms/src/Dictionaries/XDBCDictionarySource.h [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+    class Logger;
+}
+
+
+namespace DB
+{
+
+/// Allows loading dictionaries from a XDBC source via bridges
+    class XDBCDictionarySource final : public IDictionarySource
```
Comment (alesapin) [inline dbms/src/Dictionaries/XDBCDictionarySource.h]: Style

File: dbms/src/TableFunctions/ITableFunctionXDBC.cpp [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        ReadWriteBufferFromHTTP buf(columns_info_uri, Poco::Net::HTTPRequest::HTTP_POST, nullptr);
+
+        std::string columns_info;
+        readStringBinary(columns_info, buf);
+        NamesAndTypesList columns = NamesAndTypesList::parse(columns_info);
+
+        auto result = std::make_shared<StorageXDBC>(table_name, schema_name, table_name, ColumnsDescription{columns}, context, helper);
+
+        if(!result)
+            throw Exception("Failed to instantiate storage from table function " + getName());
```
Comment (alesapin) [inline dbms/src/TableFunctions/ITableFunctionXDBC.cpp]: Error Code

File: dbms/src/Storages/StorageXDBC.cpp [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+        uri = bridge_helper->getMainURI();
+    }
+
+    std::string StorageXDBC::getReadMethod() const
+    {
+        return Poco::Net::HTTPRequest::HTTP_POST;
+    }
+
+    std::vector<std::pair<std::string, std::string>> StorageXDBC::getReadURIParams(const Names & column_names,
+                                                                                   const SelectQueryInfo & /*query_info*/,
```
Comment (alesapin) [inline dbms/src/Storages/StorageXDBC.cpp]: This indent looks bad.

File: dbms/src/Common/XDBCBridgeHelper.h [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+        {
+            auto uri = createBaseURI();
+            uri.setPath(IDENTIFIER_QUOTE_HANDLER);
+            uri.addQueryParameter("connection_string", getConnectionString());
+
+            ReadWriteBufferFromHTTP buf(uri, Poco::Net::HTTPRequest::HTTP_POST, nullptr);
+            std::string character;
+            readStringBinary(character, buf);
+            if (character.length() > 1)
+                throw Exception("Failed to get quoting style from " + BridgeHelperMixin::serviceAlias());
```
Comment (alesapin) [inline dbms/src/Common/XDBCBridgeHelper.h]: ErrorCode and maybe add `character` to error message

File: dbms/programs/odbc-bridge/ColumnInfoHandler.cpp [hunk 10]
```diff
@@ -113,10 +114,22 @@ void ODBCColumnsInfoHandler::handleRequest(Poco::Net::HTTPServerRequest & reques
...
+            settings.identifier_quoting_style = IdentifierQuotingStyle::Backticks;
+        else if(identifier_quote[0] == '"')
+            settings.identifier_quoting_style = IdentifierQuotingStyle::DoubleQuotes;
+        else
+            throw Exception("Can not map quote identifier '" + identifier_quote + "' to IdentifierQuotingStyle value");
+
         select->format(settings);
         std::string query = ss.str();
 
+        std::cout << query << std::endl;
```
Comment (alesapin) [inline dbms/programs/odbc-bridge/ColumnInfoHandler.cpp]: Debug info?

File: dbms/programs/odbc-bridge/ColumnInfoHandler.cpp [hunk 11]
```diff
@@ -113,10 +114,22 @@ void ODBCColumnsInfoHandler::handleRequest(Poco::Net::HTTPServerRequest & reques
...
+
+        auto identifier_quote = getIdentifierQuote(hdbc);
+        if (identifier_quote.length() == 0)
+            settings.identifier_quoting_style = IdentifierQuotingStyle::None;
+        else if(identifier_quote[0] == '`')
+            settings.identifier_quoting_style = IdentifierQuotingStyle::Backticks;
+        else if(identifier_quote[0] == '"')
+            settings.identifier_quoting_style = IdentifierQuotingStyle::DoubleQuotes;
+        else
+            throw Exception("Can not map quote identifier '" + identifier_quote + "' to IdentifierQuotingStyle value");
```
Comment (alesapin) [inline dbms/programs/odbc-bridge/ColumnInfoHandler.cpp]: ErrorCode

**Comment by alesapin** [comment]
Style Erros: 
```dbms/src/TableFunctions/ITableFunctionXDBC.h:38: const std::string & connection_string_) const override { 
dbms/src/TableFunctions/ITableFunctionXDBC.h:55: const std::string & connection_string_) const override { 
dbms/src/Common/XDBCBridgeHelper.h:238:struct ODBCBridgeMixin {```

File: dbms/src/Common/XDBCBridgeHelper.h (hunk 5 above)
Comment (alex-krash) [reply dbms/src/Common/XDBCBridgeHelper.h]: Done.

File: dbms/programs/odbc-bridge/getIdentifierQuote.cpp (hunk 1 above)
Comment (alex-krash) [reply dbms/programs/odbc-bridge/getIdentifierQuote.cpp]: I propose to keep it consistent with bloody ODBC-interface, and not make too tricky and fragile.

**Comment by alesapin** [comment]
Build still fails with errors: https://pastebin.com/EeVHeViY

**Comment by alesapin** [comment]
Ok, thank you. Now travis build fails (you can see error in "Details" link) and two ODBC integration tests fail. 
https://pastebin.com/FuMwJTy4

**Comment by alex-krash** [comment]
Got it, thanks! Will spend more time on trying to launch integrations.

**Comment by alesapin** [comment]
Maybe it's possible to recognize error from stack trace:
```
E           QueryRuntimeException: Client failed! Return code: 232, stderr: Received exception from server (version 18.14.2):
E           Code: 1000. DB::Exception: Received from 172.18.0.3:9000. DB::Exception: Not found: dictionary.source.connection_string. Stack trace:
E          
E           0. /usr/bin/clickhouse(StackTrace::StackTrace()+0x13) [0x6873223]
E           1. /usr/bin/clickhouse(DB::TCPHandler::runImpl()+0xfbe) [0x2d6192e]
E           2. /usr/bin/clickhouse(DB::TCPHandler::run()+0x1a) [0x2d68cfa]
E           3. /usr/bin/clickhouse(Poco::Net::TCPServerConnection::start()+0xc) [0x6976acc]
E           4. /usr/bin/clickhouse(Poco::Net::TCPServerDispatcher::run()+0x118) [0x6976f98]
E           5. /usr/bin/clickhouse(Poco::PooledThread::run()+0xad) [0x6a3dfad]
E           6. /usr/bin/clickhouse(Poco::ThreadImpl::runnableEntry(void*)+0x3a) [0x6a3b72a]
E           7. /usr/bin/clickhouse() [0x705cdcf]
E           8. /lib/x86_64-linux-gnu/libpthread.so.0(+0x8184) [0x7f8568228184]
E           9. /lib/x86_64-linux-gnu/libc.so.6(clone+0x6d) [0x7f8567a4703d]
```
Intergration tests instruction https://github.com/yandex/ClickHouse/blob/master/dbms/tests/integration/README.md

**Comment by alexey-milovidov** [comment]
> Nullability support abscence. For JDBC bridge I've implemented it (see below), maybe we should implement it for ODBC as well?

We want to add the support for Nullable types in ODBC and MySQL storages.

The logic will be, for `MySQL` and `ODBC` table engine:
- convert NULL to default value, if it is not Nullable in ClickHouse table;
- read NULL value if it is Nullable in ClickHouse table.
For `mysql` and `odbc` table functions:
- introduce a setting to use Nullable types (by default) or to convert everything to non-Nullable (current behaviour).

**Comment by alexey-milovidov** [comment]
As I understand, it is disabled right now.
Nevertheless, I've merged this PR to make further development easier.

Things to do:
- enable this feature;
- prepare a manual on how to start using it after installing ClickHouse:
-- how to install JDK;
-- how to install JDBC bridge;
-- how to install example database with JDBC driver (maybe it will be ClickHouse itself);
- add integration test.

**Comment by alexey-milovidov** [comment]
We should think how to make this feature available for users with absolute minimum installation steps.
- create a jar with all dependencies inside?
- package it in selfcontained application (is it possible)?
- add it to the `clickhouse-common-static` package?

**Comment by alexey-milovidov** [comment]
> We want to add the support for Nullable types in ODBC and MySQL storages.
The logic will be, for MySQL and ODBC table engine:
    convert NULL to default value, if it is not Nullable in ClickHouse table;
    read NULL value if it is Nullable in ClickHouse table.

This is implemented in  #3362

> For mysql and odbc table functions:
    introduce a setting to use Nullable types (by default) or to convert everything to non-Nullable (current behaviour).

This is to do.

**Pull request merged**